six/modules/c++/six.sicd/source/SCPCOA.cpp

```diff
@@ -59,3 +67,314 @@ bool SCPCOA::operator==(const SCPCOA& rhs) const
+    {
+        sideOfTrack = derivedSideOfTrack(geoData);
+    }
+    if (Init::isUndefined<double>(slantRange))
+    {
+        slantRange = derivedSlantRange(scp);
+    }
+    if (Init::isUndefined<double>(groundRange))
+    {
+        groundRange = derivedGroundRange(scp);
+    }
+    if (Init::isUndefined<double>(dopplerConeAngle))
+    {
+        dopplerConeAngle = derivedDopplerConeAngle(scp);
+    }
+    if (Init::isUndefined<double>(grazeAngle))
+    {
+        grazeAngle = derivedGrazeAngle(scp);
+    }
+    if (Init::isUndefined<double>(incidenceAngle))
+    {
+        incidenceAngle = derivedIncidenceAngle(scp);
+    }
+    if (Init::isUndefined<double>(twistAngle))
+    {
+        twistAngle = derivedTwistAngle(scp);
+    }
+    if (Init::isUndefined<double>(slopeAngle))
+    {
+        slopeAngle = derivedSlopeAngle(scp);
+    }
+    if (Init::isUndefined<double>(azimAngle))
+    {
+        azimAngle = derivedAzimAngle(scp);
+    }
+    if (Init::isUndefined<double>(layoverAngle))
+    {
+        layoverAngle = derivedLayoverAngle(scp);
+    }
+}
+
+Vector3 SCPCOA::uLOS(const Vector3& scp) const
+{
+    return (scp - arpPos).unit();
+}
+
+Vector3 SCPCOA::left() const
+{
+    return cross(arpPos.unit(), arpVel.unit());
+}
+
+int SCPCOA::look(const Vector3& scp) const
+{
+    return six::sicd::Utilities::sign(uLOS(scp).dot(left()));
+}
+
+std::vector<Vector3> SCPCOA::derivedArpVectors(const Position& position) const
+{
+    std::vector<Vector3> vectors(3);
+
+    for (size_t ii = 0; ii < 3; ++ii)
+    {
+        Poly1D posPoly = sicd::Utilities::nPoly(position.arpPoly, ii);
+        Poly1D velPoly = posPoly.derivative();
+        Poly1D accPoly = velPoly.derivative();
+
+        vectors[0][ii] = posPoly(scpTime);
+        vectors[1][ii] = velPoly(scpTime);
+        vectors[2][ii] = accPoly(scpTime);
+    }
+    return vectors;
+}
+
+SideOfTrackType SCPCOA::derivedSideOfTrack(const GeoData& geoData) const
+{
+    return look(geoData.scp.ecf) < 0 ? SideOfTrackType::RIGHT :
+        SideOfTrackType::LEFT;
+}
+
+bool SCPCOA::validate(const GeoData& geoData,
+        const Grid& grid,
+        const Position& position,
+        logging::Logger& log)
+{
+    std::ostringstream messageBuilder;
+    bool valid = true;
+    const Vector3& scp = geoData.scp.ecf;
+    // 2.7.1
+    // SCPTime has stricter tolerance because everything else depends on it
+    if (std::abs(derivedSCPTime(grid) - scpTime) >
+            std::numeric_limits<double>::epsilon())
+    {
+        messageBuilder.str("");
+        messageBuilder << mScpcoaInconsistentString << "\n" <<
+            "SCPCOA.scpTime: " << scpTime << std::endl <<
+            "Derived scpTime: " << derivedSCPTime(grid) << std::endl;
+        log.error(messageBuilder.str());
+        valid = false;
+    }
+
+    std::vector<Vector3> arpVectors = derivedArpVectors(position);
+    // 2.7.2
+    valid = valid && compareFields(arpPos, arpVectors[0], "arpPos", log);
```

Comment:
OK so this may suck in terms of a refactor it would require, but it just occurred to me... if we want to validate a SICD, we probably want the validator to tell us ALL the stuff that's wrong with it.  The way checks like this are structured, if `valid` is false by the time we get here, `compareFields()` isn't going to get called because the first half of the conditional is already false.  One way to refactor would be to have some of these helper functions take in `valid` by non-const ref and update it.  For convenience, they could also return the value.  That is, it could be something like...

```
bool myComparison(int whateverOtherValues, bool& valid)
{
    bool myCompIsValid = true;

    if (whateverCondition)
    {
        myCompIsValid = false;
    }

    valid &=myCompIsValid;
    return myCompIsValid;
}
```